darkcluster/src/main/java/com/linkedin/darkcluster/impl/ConstantQpsDarkClusterStrategy.java

```diff
@@ -0,0 +1,132 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.darkcluster.impl;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.common.util.None;
+import com.linkedin.darkcluster.util.GuaranteedRateLimiter;
+import javax.annotation.Nonnull;
+
+import com.linkedin.common.util.Notifier;
+import com.linkedin.d2.DarkClusterConfig;
+import com.linkedin.d2.balancer.ServiceUnavailableException;
+import com.linkedin.d2.balancer.util.ClusterInfoProvider;
+import com.linkedin.darkcluster.api.BaseDarkClusterDispatcher;
+import com.linkedin.darkcluster.api.DarkClusterStrategy;
+import com.linkedin.r2.message.RequestContext;
+import com.linkedin.r2.message.rest.RestRequest;
+
+/**
+ * ConstantQpsDarkClusterStrategy figures out how many dark requests to send. The high level goal of this strategy is to
+ * keep the incoming QPS per dark host constant.
+ *
+ * It uses the {@link ClusterInfoProvider} to determine the number of instances in both the source and target cluster,
+ * and uses that to calculate the number of requests to send in order to make the QPS per dark host constant and equal
+ * to a specified value, assuming all hosts in the source cluster send traffic.
+ */
+public class ConstantQpsDarkClusterStrategy implements DarkClusterStrategy {
+  private final String _originalClusterName;
+  private final String _darkClusterName;
+  private final Integer _darkClusterPerHostInboundTargetRate;
+  private final BaseDarkClusterDispatcher _baseDarkClusterDispatcher;
+  private final Notifier _notifier;
+  private final ClusterInfoProvider _clusterInfoProvider;
+  private final GuaranteedRateLimiter _rateLimiter;
+
+  public ConstantQpsDarkClusterStrategy(@Nonnull String originalClusterName, @Nonnull String darkClusterName,
+      @Nonnull Integer darkClusterPerHostInboundTargetRate, @Nonnull BaseDarkClusterDispatcher baseDarkClusterDispatcher,
+      @Nonnull Notifier notifier, @Nonnull ClusterInfoProvider clusterInfoProvider, @Nonnull GuaranteedRateLimiter rateLimiter) {
+    _originalClusterName = originalClusterName;
+    _darkClusterName = darkClusterName;
+    _darkClusterPerHostInboundTargetRate = darkClusterPerHostInboundTargetRate;
+    _baseDarkClusterDispatcher = baseDarkClusterDispatcher;
+    _notifier = notifier;
+    _clusterInfoProvider = clusterInfoProvider;
+    _rateLimiter = rateLimiter;
+  }
+
+  @Override
+  public boolean handleRequest(RestRequest originalRequest, RestRequest darkRequest, RequestContext requestContext) {
+    float sendRate = getPerHostSendRate();
+    _rateLimiter.setRate(sendRate, 1000, 1);
```

Comment:
Let's avoid using magic numbers like 1000 and 1 here. For 1000, it'd better to use something like `private static final long ONE_SECOND_PERIOD = TimeUnit.SECONDS.toMillis(1);`. For the burst parameter, I think we might want to use unlimited burst here, something like, `private static final int UNLIMITED_BURST = Integer.MAX_VALUE;`

Reply:
sounds good

Reply:
Actually, on a second thought, we might want to spread out the requests during the 1 second. Given how burst is calculated (https://github.com/linkedin/rest.li/blob/7372bc76f7ce78007ecb24ca80f17399e5246e13/r2-core/src/main/java/com/linkedin/r2/transport/http/client/ratelimiter/Rate.java#L41), we might want to use the formula `burst = Math.ceil(sendRate / 1000)`. So for example, when QPS is < 1000, burst = 1. So it will be able to spread each request evenly during the 1 second. But when QPS is 2000, burst = 2, meaning that 2 requests will be issued every 1 ms.

Reply:
Next draft will include this